src/Service/Resolvers/ArrayPoolWriter.cs

```diff
@@ -0,0 +1,189 @@
+            throw new ArgumentOutOfRangeException(nameof(count));
+        }
+        
+        if(count > _capacity)
+        {
+            throw new ArgumentOutOfRangeException(nameof(count), count, "Cannot advance past the end of the buffer.");
+        }
+        
+        _start += count;
+        _capacity -= count;
+    }
+
+    /// <summary>
+    /// Gets a <see cref="Memory{T}"/> to write to.
+    /// </summary>
+    /// <param name="sizeHint">
+    /// The minimum size of the returned <see cref="Memory{T}"/>.
+    /// </param>
+    /// <returns>
+    /// A <see cref="Memory{T}"/> to write to.
+    /// </returns>
+    /// <exception cref="ArgumentOutOfRangeException">
+    /// Thrown if <paramref name="sizeHint"/> is negative.
+    /// </exception>
+    public Memory<byte> GetMemory(int sizeHint = 0)
+    {
+        if(_disposed)
+        {
+            throw new ObjectDisposedException(nameof(ArrayPoolWriter));
+        }
+        
+        if(sizeHint < 0)
+        {
+            throw new ArgumentOutOfRangeException(nameof(sizeHint));
+        }
+        
+        int size = sizeHint < 1 ? _initialBufferSize : sizeHint;
+        EnsureBufferCapacity(size);
+        return _buffer.AsMemory().Slice(_start, size);
+    }
+
+    /// <summary>
+    /// Gets a <see cref="Span{T}"/> to write to.
+    /// </summary>
+    /// <param name="sizeHint">
+    /// The minimum size of the returned <see cref="Span{T}"/>.
+    /// </param>
+    /// <returns>
+    /// A <see cref="Span{T}"/> to write to.
+    /// </returns>
+    /// <exception cref="ArgumentOutOfRangeException">
+    /// Thrown if <paramref name="sizeHint"/> is negative.
+    /// </exception>
+    public Span<byte> GetSpan(int sizeHint = 0)
+    {
+        if(_disposed)
+        {
+            throw new ObjectDisposedException(nameof(ArrayPoolWriter));
+        }
+        
+        if(sizeHint < 0)
+        {
+            throw new ArgumentOutOfRangeException(nameof(sizeHint));
+        }
+        
+        int size = sizeHint < 1 ? _initialBufferSize : sizeHint;
+        EnsureBufferCapacity(size);
+        return _buffer.AsSpan().Slice(_start, size);
+    }
+
+    /// <summary>
+    /// Ensures that the internal buffer has the needed capacity.
+    /// </summary>
+    /// <param name="neededCapacity">
+    /// The needed capacity on the internal buffer.
+    /// </param>
+    private void EnsureBufferCapacity(int neededCapacity)
+    {
+        // check if we have enough capacity available on the buffer.
+        if (_capacity < neededCapacity)
+        {
+            // if we need to expand the buffer we first capture the original buffer.
+            byte[] buffer = _buffer;
+            
+            // next we determine the new size of the buffer, we at least double the size to avoid
+            // expanding the buffer too often.
+            int newSize = buffer.Length * 2;
+            
+            // if that new buffer size is not enough to satisfy the needed capacity
```

Comment:
the comment here implies you meant to check neededCapacity to newSize

But I see the comparison happening here is still with the old buffer length - note buffer.length has not been doubled yet, its stored in `newSize`

Am I missing something?

should this comparison be: 
(neededCapacity > newSize)

Reply:
You are correct ... I will fix that.

Reply:
I have fixed this issue and added some tests.